### PR TITLE
New StaticProxyFactoryFactory

### PR DESCRIPTION
**Pull request by hazzik**

### CODE REVIEW

**Comment by hazzik** [comment]
At the current state of the PR, performance improvement is 16.8% on the data from #1446 


[Benchmarks.zip](https://github.com/nhibernate/nhibernate-core/files/1495803/Benchmarks.zip)

File: src/NHibernate/Proxy/Poco/BasicLazyInitializer.cs [hunk 1]
```diff
@@ -81,10 +81,6 @@ public virtual object Invoke(MethodInfo method, object[] args, object proxy)
 				{
 					return null;
 				}
-				else if ("get_HibernateLazyInitializer".Equals(methodName))
```
Comment (hazzik) [inline src/NHibernate/Proxy/Poco/BasicLazyInitializer.cs]: We, probably, need to revert this change if we want to support [legacy] proxy factories.

File: src/NHibernate/Proxy/DynamicProxy/NHibernateProxyImplementor.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		public static void ImplementProxy(TypeBuilder typeBuilder, FieldInfo interceptor)
+		{
+			// Implement the getter
+			var getMethod = typeBuilder.DefineMethod($"{NHibernateProxyType.FullName}.get_{HibernateLazyInitializerFieldName}", InterceptorMethodsAttributes, CallingConventions.HasThis, ReturnType, System.Type.EmptyTypes);
+			getMethod.SetImplementationFlags(MethodImplAttributes.Managed | MethodImplAttributes.IL);
+
+			var IL = getMethod.GetILGenerator();
+
+			// This is equivalent to:
+			// get { return (INHibernateProxy)__interceptor; }
```
Comment (fredericDelaporte) [inline src/NHibernate/Proxy/DynamicProxy/NHibernateProxyImplementor.cs]: Comment should be `// get { return (ILazyInitializer)__interceptor; }`

File: src/NHibernate/Proxy/Poco/BasicLazyInitializer.cs (hunk 1 above)
Comment (fredericDelaporte) [reply src/NHibernate/Proxy/Poco/BasicLazyInitializer.cs]: Yes, and it looks to me there is no easy way to obsolete this code path: I do not think we can know there if the property implementation is itself proxied or not.

File: src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs [hunk 3]
```diff
@@ -33,15 +30,12 @@ internal class DefaultMethodEmitter : IMethodBodyEmitter
...
 
-		private static readonly PropertyInfo interceptorProperty = typeof (IProxy).GetProperty("Interceptor");
-
 		private static readonly ConstructorInfo notImplementedConstructor = typeof(NotImplementedException).GetConstructor(new System.Type[0]);
 
 		private readonly IArgumentHandler _argumentHandler;
 
 		static DefaultMethodEmitter()
 		{
-			getInterceptor = interceptorProperty.GetGetMethod();
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs]: I didn't notice that it still there.

**Comment by hazzik** [comment]
Ok, I've finally come up with what I want to achieve with this PR. As all method implementations are known at the compile time, it would much beneficial to generate a "static" proxies. This means that there will be no interceptors involved. This will allow removing the unnecessary overhead of a heavy reflection manipulations.  The public classes and interfaces from DynamicProxy would be obsoleted in 5.1 and removed in 6.0. We would also need to update NHibernateProxyGenerator (NHPG) to use the new approach.

**Comment by hazzik** [comment]
Finally was able to run a benchmark. With the current state of the PR memore usage has dropped to
> NHibernate v5.0.0.0 (v5.0.3.0) : 228 951 KB (234 446 840 bytes) (more than half of 5.0.3)

And execution time has fell for about 30% (2s vs 2.6s on my laptop)

**Comment by hazzik** [comment]
Final number:
>241 491 KB

File: src/NHibernate/Proxy/AbstractProxyFactory.cs [hunk 4]
```diff
@@ -10,8 +11,9 @@ namespace NHibernate.Proxy
 	/// <summary>
 	/// Convenient common implementation for ProxyFactory
 	/// </summary>
-	public abstract class AbstractProxyFactory: IProxyFactory
+	public abstract class AbstractProxyFactory: IProxyFactory, ISerializable
```
Comment (hazzik) [inline src/NHibernate/Proxy/AbstractProxyFactory.cs]: The easiest (and I think the _correct_) way to serialize a proxy was to serialize its proxy factory.

File: src/NHibernate/Proxy/DefaultProxyFactory.cs [hunk 5]
```diff
@@ -29,10 +39,23 @@ public override INHibernateProxy GetProxy(object id, ISessionImplementor session
...
 		public override object GetFieldInterceptionProxy(object instanceToWrap)
 		{
-			var interceptor = new DefaultDynamicLazyFieldInterceptor();
-			return factory.CreateProxy(PersistentClass, interceptor, new[] { typeof(IFieldInterceptorAccessor) });
+			var factory = new ProxyFactory();
+			var proxyType = factory.CreateProxyType(PersistentClass, typeof(IFieldInterceptorAccessor));
+			var result = Activator.CreateInstance(proxyType);
+			var proxy = (IProxy) result;
+			proxy.Interceptor = new DefaultDynamicLazyFieldInterceptor();
+			return result;
```
Comment (hazzik) [inline src/NHibernate/Proxy/DefaultProxyFactory.cs]: Changes to this method need to be reverted.

File: src/NHibernate/Proxy/AbstractProxyFactory.cs (hunk 4 above)
Comment (hazzik) [inline src/NHibernate/Proxy/AbstractProxyFactory.cs]: The easiest (and I think the _correct_) way to serialize a proxy was to serialize its proxy factory.

File: src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs]: This is going to be reverted.

File: src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		NHibernateProxyObjectReference(SerializationInfo info, StreamingContext context)
+		{
+			_proxyFactory = (IProxyFactory) info.GetValue(nameof(_proxyFactory), typeof(IProxyFactory));
+			_identifier = info.GetValue(nameof(_identifier), typeof(object));
+		}
+
+		[SecurityCritical]
+		public object GetRealObject(StreamingContext context)
+		{
+			return _proxyFactory.GetProxy(_identifier, null);
```
Comment (hazzik) [inline src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs]: `null` here is for `ISessionImplementation`. On deserialization, the `PersistanceContext` will re-associate the session with the proxy. The proxy does support having a `null` session.

**Comment by gliljas** [comment]
@hazzik Well done! I will try to find some time to review it properly this week (I sing in a choir, and it's Christmas time, so like six rehearsals this week.. :) )

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+				{
+					typeof(IProxyFactory),
+					typeof(object),
+				});
+			IL.Emit(OpCodes.Newobj, constructor);
+
+			IL.Emit(OpCodes.Ldarg_1);
+			IL.Emit(OpCodes.Ldarg_2);
+
+			IL.Emit(OpCodes.Callvirt, typeof(ISerializable).GetMethod(nameof(ISerializable.GetObjectData)));
```
Comment (fredericDelaporte) [inline src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: So we have now two different ways of implementing the an `IObjectReference` with proxies, since the one for `DynamicProxy.ProxyFactory` does not delegate the `GetObjectData` to `DynamicProxy.ProxyObjectReference`. I prefer the pattern you have done for `NHibernateProxyObjectReference`, but then we should align `ProxyObjectReference` on the same pattern.

File: src/NHibernate.Test/NHSpecificTest/NH3954/EqualsFixture.cs [hunk 8]
```diff
@@ -48,9 +49,9 @@ public void TypeInequality()
 		[Test]
 		public void InterfaceEquality()
 		{
-			var entry1 = new ProxyCacheEntry(typeof(Entity1), new[] { typeof(INHibernateProxy), typeof(IProxy) });
+			var entry1 = new ProxyCacheEntry(typeof(Entity1), new[] { typeof(INHibernateProxy), typeof(IFieldInterceptorAccessor) });
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate.Test/NHSpecificTest/NH3954/EqualsFixture.cs]: It’s not “not fully done” it’s simply “not done”. Yeah, going to rebase on #1454 then cleanup

File: src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs (hunk 6 above)
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs]: > it may be a breaking changed for a custom `IProxyFactory`

In this codepath there will be no custom `IProxyFactory`. 

The LazyInitializer underneath does support accepting `null`.

So, answer is no.

File: src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/DefaultMethodEmitter.cs]: Yes

File: src/NHibernate/Proxy/AbstractProxyFactory.cs (hunk 4 above)
Comment (hazzik) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: Maybe...

File: src/NHibernate/Proxy/AbstractProxyFactory.cs [hunk 9]
```diff
@@ -25,6 +27,21 @@ protected bool IsClassProxy
...
 
+		protected AbstractProxyFactory()
+		{
+		}
+
+		protected AbstractProxyFactory(SerializationInfo info, StreamingContext context)
+		{
+			EntityName = (string) info.GetValue(nameof(EntityName), typeof(string));
+			PersistentClass = (System.Type) info.GetValue(nameof(PersistentClass), typeof(System.Type));
+			Interfaces = (System.Type[]) info.GetValue(nameof(Interfaces), typeof(System.Type[]));
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: I don’t think that proxies would even work in (all) netstandard as Reflection.Emit is a netstandard 1.x + platforms rather than a pure netstandard. 

#1425 would not really help here as I’m seraializing a `MethodInfo` as well.

File: src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
+using System;
+using System.Runtime.Serialization;
+using System.Security;
+
+namespace NHibernate.Proxy.DynamicProxy
+{
+	[Serializable]
+	public sealed class NHibernateProxyObjectReference : IObjectReference, ISerializable
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/NHibernateProxyObjectReference.cs]: True

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs (hunk 7 above)
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Not going to bother with DynamicProxy namespace. Going to obsolete it after converting lazy properties proxies to the same approach.

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+		}
+
+		static void ImplementGetObjectData(System.Type baseType, IReadOnlyCollection<System.Type> baseInterfaces, TypeBuilder typeBuilder, FieldBuilder lazyInitializerField, FieldBuilder proxyFactoryField)
+		{
+			const MethodAttributes attributes = MethodAttributes.Public | MethodAttributes.HideBySig |
+			                                    MethodAttributes.Virtual;
+			var parameterTypes = new[] {typeof (SerializationInfo), typeof (StreamingContext)};
+
+			MethodBuilder methodBuilder =
+				typeBuilder.DefineMethod("GetObjectData", attributes, typeof (void), parameterTypes);
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Yes

File: src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs [hunk 12]
```diff
@@ -257,7 +240,7 @@ private static void ImplementGetObjectData(System.Type baseType, System.Type[] b
 			IL.Emit(OpCodes.Ret);
 		}
 
-		private static void DefineSerializationConstructor(TypeBuilder typeBuilder, FieldInfo interceptorField, ConstructorBuilder defaultConstructor)
+		internal static void DefineSerializationConstructor(TypeBuilder typeBuilder, FieldInfo interceptorField, ConstructorBuilder defaultConstructor)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs]: I don't want to touch/fix/change DynamicProxy namespace other than for major bugs. It's working, so let it be this way. The namespace itself is going to be retired in 6.0

File: src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs [hunk 13]
```diff
@@ -206,7 +189,7 @@ private static ConstructorBuilder DefineConstructor(TypeBuilder typeBuilder, Sys
 			return constructor;
 		}
 
-		private static void ImplementGetObjectData(System.Type baseType, System.Type[] baseInterfaces, TypeBuilder typeBuilder, FieldInfo interceptorField)
+		internal static void ImplementGetObjectData(System.Type baseType, IReadOnlyCollection<System.Type> baseInterfaces, TypeBuilder typeBuilder, FieldInfo interceptorField)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs]: We should, but I don't want to touch the DynamicProxy namespace

File: src/NHibernate/Proxy/AbstractProxyFactory.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ngbrown) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: `MethodInfo` in covered in #1425.  I don't think NHibernate on .netstandard works without things like DynamicProxies.  The target isn't iOS (which requires static compilation), but the targets able to run .Net Core.

[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: > Update: Is this pull request going to provide a compile-time proxy generation ability?

It looks to me as being the purpose of https://github.com/nhibernate/NHibernate.ProxyGenerators. I do not expect this to change. Moreover Alexander [had mentioned](https://github.com/nhibernate/nhibernate-core/pull/1451#issuecomment-346935633) this project will need an update after this PR.

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: I have plans to include proxy generators into the core (as a tool + msbuild targets) to provide support for compile time proxies for the platforms which do not support runtime proxies. But not in this PR, and, most likely, not in this milestone.

**Comment by hazzik** [comment]
I've cleaned the PR addressing some of the review comments.

**Comment by hazzik** [comment]
I have minimized changes to the old code.

**Comment by hazzik** [comment]
Now it's implemented as a completely new feature.

**Comment by fredericDelaporte** [comment]
I am not forgetting it and was intending to look it again today anyway.

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+			var IL = methodBuilder.GetILGenerator();
+			//LocalBuilder proxyBaseType = IL.DeclareLocal(typeof(Type));
+
+			// info.SetType(typeof(NHibernateProxyObjectReference));
+			IL.Emit(OpCodes.Ldarg_1);
+			IL.Emit(OpCodes.Ldtoken, typeof (NHibernateProxyObjectReference));
+			IL.Emit(OpCodes.Call, ReflectionCache.TypeMethods.GetTypeFromHandle);
+			IL.Emit(OpCodes.Callvirt, ProxyFactory.setType);
+
+			//this.__proxyInfo
```
Comment (fredericDelaporte) [inline src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: I would add before:
```c#
// (new NHibernateProxyObjectReference(__proxyInfo, __lazyInitializer.Identifier)).GetObjectDataMethod(info, context);
```

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+
+			typeBuilder.DefineMethodOverride(methodOverride, method);
+		}
+
+		static void ImplementSetIdentifier(TypeBuilder typeBuilder, MethodInfo method, FieldInfo lazyInitializerField)
+		{
+			/*
+			 set 
+			 {
+				this.__lazyInitializer.Initialize();
```
Comment (fredericDelaporte) [inline src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Indeed should be prepended with:
```c#
if (this.__lazyInitializer == null)
	return base.<Method> (args..)
```

Same for next two methods.

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+using NHibernate.Proxy.DynamicProxy;
+using NHibernate.Type;
+using NHibernate.Util;
+
+namespace NHibernate.Proxy
+{
+	class NHibernateProxyBuilder
+	{
+		const MethodAttributes InterceptorMethodsAttributes = MethodAttributes.Private | MethodAttributes.Final | MethodAttributes.HideBySig |
+		                                                      MethodAttributes.SpecialName | MethodAttributes.NewSlot | MethodAttributes.Virtual;
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Nothing strange, actually. This is how explicit interface members defined in CLR.

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Ok, here are the explanation:

1. `Private` - this is an explicit method.
2. `Final` - equivalent to `sealed`
3. `HideBySig` - that it's will hide/be hidden by name and signature 
4. `SpecialName` - that it is a property
5. `NewSlot` - not sure, but this is how interface methods are implemented
6. `Virtual` - that it's overridable

On CIL level interface methods are indistinguishable from public abstract members. The difference between interface and abstract class are in the type declaration. In CIL these members are declared as "public hidebysig newslot abstract virtual". The methods which implement the interface are equivalent to implementations of abstract non-overridable methods: `override sealed` (CIL: `final hidebysig virtual`), with the exeption that interface implementations get a new slot (`newslot`).

See this [SharpLab](http://bit.ly/2Af2Lo6)

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+		static readonly IProxyAssemblyBuilder ProxyAssemblyBuilder = new DefaultProxyAssemblyBuilder();
+
+		static readonly ConstructorInfo SecurityCriticalAttributeConstructor = typeof(SecurityCriticalAttribute).GetConstructor(System.Type.EmptyTypes);
+		static readonly MethodInfo SerializableGetObjectDataMethod = typeof(ISerializable).GetMethod(nameof(ISerializable.GetObjectData));
+
+		readonly MethodInfo _getIdentifierMethod;
+		readonly MethodInfo _setIdentifierMethod;
+		readonly IAbstractComponentType _componentIdType;
+		readonly bool _overridesEquals;
+		public NHibernateProxyBuilder(MethodInfo getIdentifierMethod, MethodInfo setIdentifierMethod, IAbstractComponentType componentIdType, bool overridesEquals)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Ok

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+			{
+				//skip
+			}
+			else if (!_overridesEquals && method.Name == "GetHashCode" && method.GetBaseDefinition() == typeof(object).GetMethod("GetHashCode"))
+			{
+				//skip
+			}
+			else if (_componentIdType != null && _componentIdType.IsMethodOf(method))
+			{
+				ImplementCallMethodOnEmbeddedComponentId(typeBuilder, method, lazyInitializerField);
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Not sure what is that, or how it's supposed to be used. I've copied this from DefaultProxyFactory. Do you think that this path needs to be removed?

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: Ok, thanks for detailed explanation.

**Comment by hazzik** [comment]
> Shouldn't `DefaultProxyFactoryFactory`, `DefaultProxyFactory`,`DefaultLazyInitializer`, `BasicLazyInitializer` be obsoleted?

Maybe in next PR.

File: src/NHibernate/Proxy/NHibernateProxyBuilder.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Proxy/NHibernateProxyBuilder.cs]: With what I have seen on Hibernate side, it now looks to me as something partially ported, causing it to very likely never have any usage. Fully ascertaining this would require a bit more digging on Hibernate side to see in which cases it serves.
I am always for removing useless code, but since it does not cause any harm, it also not really worth it to dig more for checking it can really be removed without any risk.

So if you are not sure about it either, it is going to stay for now.

File: src/NHibernate/Bytecode/StaticProxyFactoryFactory.cs [hunk 19]
```diff
@@ -0,0 +1,17 @@
+using NHibernate.Proxy;
+
+namespace NHibernate.Bytecode
+{
+	public class StaticProxyFactoryFactory : IProxyFactoryFactory
+	{
+		internal static StaticProxyFactoryFactory Instance = new StaticProxyFactoryFactory();
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Bytecode/StaticProxyFactoryFactory.cs]: This is a small memory optimization. I don't want to enforce the singleton pattern here. And we cannot enforce it here, as `StaticProxyFactoryFactory` can be manually configured as a proxy factory and will require the public constructor to be creatable by reflection.

[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Bytecode/StaticProxyFactoryFactory.cs]: Yes, forgot to think about configuration.

File: src/NHibernate/Proxy/AbstractProxyFactory.cs (hunk 4 above)
Comment (fredericDelaporte) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: I now realize this point is a possible breaking change: static proxies are serialized without their entity state, even if they were initialized. They are always de-serialized in un-initialized state.

So, for proxies de-serialized by a session de-serialization, using them will always triggers a lazy load even if they were serialized while already being loaded. It looks to me dynamic proxies would not trigger a lazy load in that case.

For proxies directly serialized/de-serialized, with dynamic proxies their de-serialized state should be usable if they were initialized before serialization, while static proxies would be always un-initialized and un-usable because no more bound to a session.

Maybe a way to get the initializer `_target` instance should be added, for serializing it along the other proxy data, thus allowing to supply it back to the proxy on deserialization.
